data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/S3SinkWorker.java

```diff
@@ -0,0 +1,129 @@
+import org.opensearch.dataprepper.plugins.sink.accumulator.BufferTypeOptions;
+import org.opensearch.dataprepper.plugins.sink.accumulator.InMemoryBuffer;
+import org.opensearch.dataprepper.plugins.sink.accumulator.LocalFileBuffer;
+import org.opensearch.dataprepper.plugins.sink.codec.Codec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import software.amazon.awssdk.services.s3.S3Client;
+
+/**
+ * Class responsible for threshold check and instantiation of {@link BufferType}
+ */
+public class S3SinkWorker {
+
+    private static final Logger LOG = LoggerFactory.getLogger(S3SinkWorker.class);
+    private final S3Client s3Client;
+    private final S3SinkConfig s3SinkConfig;
+    private final Codec codec;
+    private final int numEvents;
+    private final ByteCount byteCapacity;
+    private final long duration;
+
+    /**
+     * @param s3Client
+     * @param s3SinkConfig
+     * @param codec
+     */
+    public S3SinkWorker(final S3Client s3Client, final S3SinkConfig s3SinkConfig, final Codec codec) {
+        this.s3Client = s3Client;
+        this.s3SinkConfig = s3SinkConfig;
+        this.codec = codec;
+        numEvents = s3SinkConfig.getThresholdOptions().getEventCount();
+        byteCapacity = s3SinkConfig.getThresholdOptions().getMaximumSize();
+        duration = s3SinkConfig.getThresholdOptions().getEventCollect().getSeconds();
+    }
+
+    /**
+     * Accumulates data from buffer and store in local-file/in-memory.
+     *
+     * @param eventQueue
+     */
+    public void bufferAccumulator(BlockingQueue<Event> eventQueue) {
+        boolean isFileUploadedToS3 = Boolean.FALSE;
+        DB eventDb = null;
+        NavigableSet<String> bufferedEventSet = null;
+        int byteCount = 0;
+        int eventCount = 0;
+        long eventCollectionDuration = 0;
+        try {
+            StopWatch watch = new StopWatch();
+            watch.start();
+            eventDb = DBMaker.memoryDB().make();
+            bufferedEventSet = eventDb.treeSet("set").serializer(Serializer.STRING).createOrOpen();
+            int data = 0;
+            while (thresholdsCheck(data, watch, byteCount)) {
+                if (!eventQueue.isEmpty()) {
+                    Event event = eventQueue.take();
+                    String jsonSerEvent = codec.parse(event);
+                    byteCount += jsonSerEvent.getBytes().length;
+                    bufferedEventSet.add(codec.parse(event).concat(System.lineSeparator()));
+                    eventCount++;
+                    data++;
+                    eventCollectionDuration = watch.getTime(TimeUnit.SECONDS);
+                }
+            }
+            eventDb.commit();
+            LOG.info(
+                    "Snapshot info : Byte_capacity = {} Bytes, Event_count = {} Records & Event_collection_duration = {} Sec",
+                    byteCount, eventCount, eventCollectionDuration);
+
+            if (s3SinkConfig.getBufferType().equals(BufferTypeOptions.LOCALFILE)) {

```

Comment:
Every time the Data Prepper pipeline sends events to the sink, it calls `doOutput` on the `S3Sink` class. This in turn, adds events to a `BlockingQueue` and then calls this method.

This method is then taking events from the `BlockingQueue`, adds them to a new `NavigableSet`. After that it, it creates the buffer (local-file or in-memory). Then it calls the "accumulate" methods.

These methods then immediately send to S3. So there is no buffer accumulation. The idea behind the accumulation is to keep data from multiple calls to `doOutput`. Otherwise, the user may get very small S3 objects.

Reply:
These changes will be addressed as incremental changes, and they will be available in a separate PR.